### PR TITLE
Mfkey qol improvements

### PR DESCRIPTION
**Pull request by GameTec-live**
mfkey32 now with saving! Save to file, New Dictionary or Existing Dictionary with ease....

### CODE REVIEW

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 1]
```diff
@@ -68,6 +79,8 @@ class Mfkey32PageState extends State<Mfkey32Page> {
                 nr1Enc: item1.nr,
                 ar1Enc: item1.ar,
               );
+              keys.add(
+                  u64ToBytes((await recovery.mfkey32(mfkey))[0]).sublist(2, 8));
               controller.text +=
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:84]: As I said remove whole textbox

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 2]
```diff
@@ -80,6 +93,21 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
 
+  List<Uint8List> deduplicateKeys(List<Uint8List> keys) {
+    return <int, Uint8List>{
+      for (var v in keys) Object.hashAll(v): v
+    }.values.toList();
+  }
+
+  String convertUintToDictFile(List<Uint8List> keys) {
+    List<Uint8List> dekeys = deduplicateKeys(keys);
+    String filecontents = "";
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:104]: fileContents

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 3]
```diff
@@ -80,6 +93,21 @@ class Mfkey32PageState extends State<Mfkey32Page> {
     }
   }
 
+  List<Uint8List> deduplicateKeys(List<Uint8List> keys) {
+    return <int, Uint8List>{
+      for (var v in keys) Object.hashAll(v): v
+    }.values.toList();
+  }
+
+  String convertUintToDictFile(List<Uint8List> keys) {
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:102]: convertKeysToDictFile

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 4]
```diff
@@ -80,6 +93,21 @@ class Mfkey32PageState extends State<Mfkey32Page> {
     }
   }
 
+  List<Uint8List> deduplicateKeys(List<Uint8List> keys) {
+    return <int, Uint8List>{
+      for (var v in keys) Object.hashAll(v): v
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:98]: for (var key in keys)

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 5]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                              context: context,
+                              builder: (BuildContext context) => AlertDialog(
+                                    title: Text("Save recovered Keys"),
+                                    content: Text(
+                                        "Where do you want to save the recovered keys?"),
+                                    actions: [
+                                      ElevatedButton(
+                                        onPressed: () async {
+                                          String filecontents =
+                                              convertUintToDictFile(keys);
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:183]: fileContents

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 6]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                          try {
+                                            await FileSaver.instance.saveAs(
+                                                name: name,
+                                                bytes: const Utf8Encoder().convert(filecontents),
+                                                ext: 'dict',
+                                                mimeType: MimeType.other);
+                                          } on UnimplementedError catch (_) {
+                                            String? outputFile = await FilePicker.platform.saveFile(
+                                              dialogTitle: '${localizations.output_file}:',
+                                              fileName: '${name}.dict',
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:194]: .dic

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 7]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                      ElevatedButton(
+                                        onPressed: () async {
+                                          String filecontents =
+                                              convertUintToDictFile(keys);
+                                          String name = Uuid().v4();
+                                          try {
+                                            await FileSaver.instance.saveAs(
+                                                name: name,
+                                                bytes: const Utf8Encoder().convert(filecontents),
+                                                ext: 'dict',
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:189]: dic

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 8]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                        },
+                                        child: Text("Save to file"),
+                                      ),
+                                      ElevatedButton(
+                                        onPressed: () async {
+                                          await dictSelectDialog(context, deduplicateKeys(keys));
+                                          Navigator.pop(context);
+                                        },
+                                        child: Text(
+                                            "Append to existing Dictionary"),
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:211]: Add to existing dictionary

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 9]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                      ),
+                                      ElevatedButton(
+                                        onPressed: () async {
+                                          Dictionary dict = Dictionary(
+                                            name: "Recovered Keys",
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:216]: Use UID here

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 10]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                            color: Colors.blue,
+                                            keys: deduplicateKeys(keys),
+                                          );
+                                          await showDialog<String>(
+                                            context: context,
+                                            builder: (BuildContext context) {
+                                              return DictionaryEditMenu(dict: dict, isNew: true,);
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:223]: DictionaryEditMenu(dict: dict, isNew: true)

File: chameleonultragui/lib/gui/page/mfkey32.dart [hunk 11]
```diff
@@ -129,11 +157,82 @@ class Mfkey32PageState extends State<Mfkey32Page> {
...
+                                            },
+                                          );
+                                          Navigator.pop(context);
+                                        },
+                                        child: Text("Save to new Dictionary"),
```
Comment (Foxushka) [inline chameleonultragui/lib/gui/page/mfkey32.dart:228]: Save to new dictionary

**Review by Foxushka — CHANGES_REQUESTED** [review]

**Comment by Foxushka** [comment]
And add translation strings

**Comment by GameTec-live** [comment]
translation is @aruznieto s job... neither yours nor my :p

**Comment by aruznieto** [comment]
> translation is @aruznieto s job... neither yours nor my :p

For this reasson i said that I preferred to use my Crowdin project...

**Review by Akisame-AI — APPROVED** [review]
It works. 
I get a random key when reading with the phone (D3F7D3F7D3F7) but also a valid key. 
Nothing breaks so I approve.

**Pull request merged**